x/gov/keeper/deposit.go

```diff
@@ -70,6 +72,51 @@ func (keeper Keeper) AddDeposit(ctx context.Context, proposalID uint64, deposito
 		return false, errors.Wrapf(types.ErrInactiveProposal, "%d", proposalID)
 	}
 
+	params, err := keeper.Params.Get(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	minDepositAmount := proposal.GetMinDepositFromParams(params)
+	minDepositRatio, err := sdkmath.LegacyNewDecFromStr(params.GetMinDepositRatio())
+	if err != nil {
+		return false, err
+	}
+
+	// the deposit must only contain valid denoms (listed in the min deposit param)
+	if !depositAmount.DenomsSubsetOf(minDepositAmount) {
+		return false, errors.Wrapf(types.ErrInvalidDepositDenom, "deposit contains invalid denom/s %s, accepted denoms are: %s", depositAmount.Denoms(), minDepositAmount.Denoms())
+	}
+
+	// if minDepositRatio is set, the deposit must be equal or greater than minDepositAmount*minDepositRatio
+	// for at least one denom. If minDepositRatio is zero we skip this check.
+	if !minDepositRatio.IsZero() {
```

Comment:
So this changes the semantics and behavior of the current min deposit logic. The current logic is a proposal's deposit `D >= MinDeposit`. 

Now the logic is `D` must be >= `MinDepositRatio * MinDeposit`, for ANY denom, not all. If this is the case, why have `MinDepositRatio` at all? Instead, just refactor the existing `MinDeposit`.

Reply:
`MinDeposit` is `Minimum deposit for a proposal to enter voting period.` so it's not being used for checking that.

The current behavior is:
- when sending a new proposal, deposit can be almost non existent
- when sending a deposit, after the proposal is on chain, the amount deposited can be extremely little

The new behavior with this PR is that deposits must be `>= MinDepositRatio * MinDeposit` for at least one of the denoms listed in MinDeposit (both for initialDeposit and deposits made after the proposal has been posted).

**Why only one denom threshold must be met?**
This was pretty much arbitrary so it's open to discussion. But mainly because I didn't want to disrupt the current flow that much, so users can choose if they want to contribute 1 or multiple denoms to a proposal's deposit. So I think this is the most flexible approach.
If a chain wants to require 100 different coins for a proposal deposit, they can do it while at the same time they can also increase `MinDepositRatio`.

Reply:
The old logic was:

```go
	params, err := keeper.Params.Get(ctx)
	if err != nil {
		return false, err
	}
	minDepositAmount := proposal.GetMinDepositFromParams(params)
```

My point is, it's not clear why we need the ratio. In the PR body you state the following deposits are acceptable:

```
- 1stake, 0atom
- 0stake, 1atom
- 1stake, 0.1atom
- 5stake, 0atom
```

My point is that you can just set `MinDeposit = {1stake, 0.1atom}`. Then upon deposit (either initial or after), you just ensure that deposit is a subset of `MinDeposit` (both by denom and value).

Am I missing something that ratio allows us to do that this doesn't?

Reply:
The code you quoted is the minimum deposit for a proposal to get into the voting period, it's not related to the minimum amount a deposit must have which is what this PR is about